neuraxle/distributed/streaming.py

```diff
@@ -0,0 +1,167 @@
+        CustomPipelineMixin.__init__(self)
+
+        self.max_batches = max_batches
+        self.batch_size = batch_size
+
+        Pipeline.__init__(self, steps=self._initialize_steps_as_tuple(steps), cache_folder=cache_folder)
+
+    def _initialize_steps_as_tuple(self, steps):
+        if not isinstance(steps, tuple):
+            steps = [(step.name, step) for step in steps]
+
+        steps_as_tuple: NamedTupleList = []
+        for name, n_workers, step in steps:
+            wrapped_step = QueueWorker(step, n_workers=n_workers)
+            if len(steps_as_tuple) > 0:
+                previous_step: Observer = steps_as_tuple[-1][1]
+                wrapped_step.subscribe(previous_step)
+
+            steps_as_tuple.append((name, wrapped_step))
+
+        return steps_as_tuple
+
+    def _will_transform_data_container(self, data_container: DataContainer, context: ExecutionContext) -> ('BaseStep', DataContainer):
+        for name, step in self:
+            step.start(context)
+
+    def _transform_data_container(self, data_container: DataContainer, context: ExecutionContext) -> DataContainer:
+        """
+        Transform data container
+
+        :param data_container: data container to transform.
+        :param context: execution context
+        :return: data container
+        """
+        data_container_batches = data_container.convolved_1d(stride=self.batch_size, kernel_size=self.batch_size)
+        n_batches = data_container.get_n_baches(self.batch_size)
+        queue_joiner = QueueJoiner(n_batches=n_batches, max_batches=self.max_batches)
+
+        batches_observable = Observable()
+        batches_observable.subscribe(self[0])
+        self[-1].subscribe(queue_joiner)
+
+        for data_container_batch in data_container_batches:
+            queue_joiner.add_batch_in_progress()
+            batches_observable.on_next(data_container_batch)
+
+        return queue_joiner.join()
+
+    def _did_transform(self, data_container: DataContainer, context: ExecutionContext) -> DataContainer:
+        for name, step in self:
+            step.stop()
+
+        return data_container
+
+
+class QueueJoiner(Observer):
+    def __init__(self, n_batches, max_batches):
+        self.mutex = Lock()
+        self.mutex.acquire()
+        self.max_batches = max_batches
+        self.n_batches_left_to_do = n_batches
+        self.batches_in_progress = 0
+        self.result = ListDataContainer(current_ids=[], data_inputs=[], expected_outputs=[], summary_id=None)
+
+    def add_batch_in_progress(self):
+        self.batches_in_progress += 1
+
+    def on_next(self, value):
+        self.n_batches_left_to_do -= 1
+        self.batches_in_progress -= 1
+
+        self.result.concat(value)
+
+        if self.n_batches_left_to_do == 0:
+            self.mutex.release()
+
+    def join(self) -> DataContainer:
+        self.mutex.acquire()
+        return self.result
```

Comment:
with the summary id of the data container for each batch it would be possible to reorder the results properly if needed